Implement shallow cloning

vig42 opened the pull request:
Fixes #1450

```
Currently, RepoSense downloads the entire commit history of the repo
when cloning it. This results in a lot of unnecessary history being
downloaded, which increases the size of the cloned repo and slows
down the time taken to clone the repo.

Let’s implement Git’s shallow cloning functionality so as to truncate
the downloaded history thereby reducing the size of the cloned repo
and speeding up the cloning process.
```

vig42 commented:
System test is failing because the "lastModifiedDate" does not match. This is because of the shallow clone functionality - since the commit history of the cloned repo is truncated, RepoSense does not see any changes made prior to the `--since` date. As such, for lines that were last modified before the `--since` date, the "lastModifiedDate" value is incorrect.

The difference is only visible in the authorship.json when the `--last-modified-date` option is used. The actual reports generated are identical to those generated by the master branch.

Tejas2805 commented:
> The difference is only visible in the authorship.json when the --last-modified-date option is used

If I understand this correctly, it means that the `--last-modified-date` option will not work if we switch to shallow cloning?

vig42 commented:
> 
> 
> > The difference is only visible in the authorship.json when the --last-modified-date option is used
> 
> If I understand this correctly, it means that the `--last-modified-date` option will not work if we switch to shallow cloning?

Yes. With shallow cloning, the `--last-modified-date` would provide wrong values for all commits made prior to the `--shallow-since` date passed to the `git clone` command.

Tejas2805 commented:
> As I had mentioned in issue #1450, I feel that it might be better to make the shallow cloning an option in the repo config. Then we could also specify in the documentation that the --last-modified-date option does not work on repos that have this option turned on. Would that be a better way to handle this?

Yup, we can proceed with it, I think. Plus it would be good to state that shallow cloning should only be used for large repositories and not for a large number of repositories considering the overhead. 

Also, based on what I read in the issue, are we avoiding partial cloning unless the 2 cases occur? Or will the partial cloning occur irrespective of the two cases?

vig42 commented:
>
> Yup, we can proceed with it, I think. Plus it would be good to state that shallow cloning should only be used for large repositories and not for a large number of repositories considering the overhead.

Will do.

> 
> Also, based on what I read in the issue, are we avoiding partial cloning unless the 2 cases occur? Or will the partial cloning occur irrespective of the two cases?

Actually we need to perform partial cloning to detect the 2 cases in the first place, so we would have to do it regardless.

damithc commented:
@vig42 what's the verdict on costs and benefits? Can give a summary?

vig42 commented:
I have made some changes to implement this functionality as a CLI flag `--optimize-cloning` and an option in repo-config.csv. By default, the option would not be enabled, i.e. RepoSense would not use shallow cloning.

> 
> 
> @vig42 what's the verdict on costs and benefits? Can give a summary?

The benefit is the ability to significantly reduce the download size of a large repository, thereby also reducing the time taken to clone the repo. I had provided some examples in #1450 where the download size of a large repo can be reduced by more than 50%, and as much as 90%, if we only look at the most recent 6 months of commits.

The first issue is that there is some overhead since we need to perform a partial clone to check for edge cases. While this overhead is insignificant relative to the massive time savings on large repos, it is quite noticeable on small repos and especially results in significantly slower performance when run on a large number of small repos.

However, given the current implementation as a config option, which can be turned on or off for each repo, this issue is not really relevant. As @Tejas2805 suggested, I have specified in the documentation that the option should not be used for smaller repos. As such, users should not experience the overhead.

The second issue is that partial cloning - which is needed to check for the edge cases - may not be compatible with all remote repositories. While it is supported by GitHub and GitLab, BitBucket notably does not support this functionality yet. If that is the case, then RepoSense will fall back to the usual method of cloning, which would also result in some overhead.

damithc commented:
Thanks for the summary @vig42 
I wonder if the flag name is too general e.g., what if we introduce other ways to optimize cloning in future? Perhaps make it more specific? e.g., `--partial-cloning`?

damithc commented on docs/ug/cli.md:
```diff
@@ -107,6 +107,17 @@ The last modified dates will be in the same timezone specified with the `--timez
 
+### `--optimize-cloning`, `-m`
+
+**`--optimize-cloning`**: Clones repositories using Git's shallow cloning functionality, which can significantly reduce the time taken to clone large repositories. However, the flag should not be used for smaller repositories. Note: this flag is incompatible with the `--last-modified-date`.
```
You might want to explain why it should not be used for small repos. also, what's considered a 'small' repo?

vig42 replied:
> 
> 
> You might want to explain why it should not be used for small repos.

Noted, will add a point about the overhead.

>
>
> also, what's considered a 'small' repo?

The exact threshold isn't very clear, since the reduction in size due to shallow cloning varies depending on when the commits were made relative to the time window of the shallow clone. But I think it is safe to say that shallow cloning would not be very helpful if the size of the output of the bare clone (i.e. the ".git" file) is less than 1GB.

damithc replied:
> The exact threshold isn't very clear, since the reduction in size due to shallow cloning varies depending on when the commits were made relative to the time window of the shallow clone. But I think it is safe to say that shallow cloning would not be very helpful if the size of the output of the bare clone (i.e. the ".git" file) is less than 1GB.

I see. You may want to mention this in the documentation as well.

damithc commented on docs/ug/configFiles.md:
```diff
@@ -38,6 +38,7 @@ Ignore Glob List<sup>*+</sup> | The list of file path globs to ignore during ana
 Ignore standalone config | To ignore the standalone config file (if any) in the target repository, enter **`yes`**. If the cell is empty, the standalone config file in the repo (if any) will take precedence over configurations provided in the csv files.
 Ignore Commit List<sup>*+</sup> | The list of commits to ignore during analysis. For accurate results, the commits should be provided with their full hash. Additionally, a range of commits can be specified using the `..` notation, e.g., `abc123..def456` (both inclusive).
 Ignore Authors List<sup>*+</sup> | The list of authors to ignore during analysis. Authors should be specified by their [Git Author Name](#a-note-about-git-author-name).
+Optimize Cloning | Enter **`yes`** to clone the repository using git's shallow cloning functionality. This option can significantly reduce the time taken to clone large repositories. However, the option should ideally be disabled for smaller repositories. Note: this is incompatible with the "Ignore standalone config" option.
```
Is this a breaking change? i.e., will this break config files in the old format? In that case, see if you can make it backward compatible.

vig42 replied:
> 
> 
> Is this a breaking change? i.e., will this break config files in the old format? In that case, see if you can make it backward compatible.

As it stands, it is a breaking change, because the CsvParser validates the repo-config.csv file by checking the number of columns. The expected number of columns in the csv file is now 8 rather than 7, so the CsvParser would throw an exception if there are only 7 columns.

We can make it backward compatible by loosening the restriction on the number of columns - we can set a minimum and maximum number of columns in the repo-config.csv file, so we would accept a csv with at least 7 and at most 8 columns. Since the new column is the rightmost, the position of the other columns in the repo-config.csv would all still be the same.

Alternatively, we could instead make CsvParser validate the csv file by checking the column headers. We could make it so that this particular column header is unnecessary, so it just sets a default value of "false". This would also handle cases where a user makes a config csv with the columns in a different order, since the CsvParser currently only cares about the position of the columns (e.g. "File Formats" is assumed to be in column 2) and ignores the headers entirely.

damithc replied:
> Alternatively, we could instead make CsvParser validate the csv file by checking the column headers. We could make it so that this particular column header is unnecessary, so it just sets a default value of "false". This would also handle cases where a user makes a config csv with the columns in a different order, since the CsvParser currently only cares about the position of the columns (e.g. "File Formats" is assumed to be in column 2) and ignores the headers entirely.

I like this idea. In fact, making some columns optional and not caring about the order of columns would be even better. This may have to be a separate PR. But unless it is done, I'm afraid we can't merge this PR because it will break currently deployed dashboards that use the latest build of RepoSense every time the report is updated.

vig42 replied:
> 
> I like this idea. In fact, making some columns optional and not caring about the order of columns would be even better. This may have to be a separate PR. But unless it is done, I'm afraid we can't merge this PR because it will break currently deployed dashboards that use the latest build of RepoSense every time the report is updated.

Understood, I will make a separate PR with that change. It would also make it easier to add other configuration options going forward without breaking backward compatibility.

vig42 replied:
> 
> 
> > Alternatively, we could instead make CsvParser validate the csv file by checking the column headers. We could make it so that this particular column header is unnecessary, so it just sets a default value of "false". This would also handle cases where a user makes a config csv with the columns in a different order, since the CsvParser currently only cares about the position of the columns (e.g. "File Formats" is assumed to be in column 2) and ignores the headers entirely.
> 
> I like this idea. In fact, making some columns optional and not caring about the order of columns would be even better. This may have to be a separate PR. But unless it is done, I'm afraid we can't merge this PR because it will break currently deployed dashboards that use the latest build of RepoSense every time the report is updated.

Prof, I have implemented this change and ran into an issue while testing. Unfortunately, a bunch of the JUnit tests fail since the config files that are being used for testing themselves do not follow the column header formats exactly.

- AuthorConfigParserTest: test config files have the column header "Author's Email", while the user guide specifies that it should be "Author's Emails"
- GroupConfigParserTest: test config files have the column header "Group label" while the user guide specifies that it should be "Group Name"
- RepoConfigParserTest: test config files have the column header "Ignore Commits List" while the user guide specifies that it should be "Ignore Commit List"

These are relatively small discrepancies. Right now I am using case-insensitive string matching to check the headers, but we can account for these differences quite easily using regex instead. Otherwise, we can just change the headers to match the user guide.

However, my worry is that users may also be using config files that do not match the user guide specification, which would mean that we would still end up breaking backward compatibility.

damithc replied:
> Prof, I have implemented this change and ran into an issue while testing. Unfortunately, a bunch of the JUnit tests fail since the config files that are being used for testing themselves do not follow the column header formats exactly.
> 
> * AuthorConfigParserTest: test config files have the column header "Author's Email", while the user guide specifies that it should be "Author's Emails"
> * GroupConfigParserTest: test config files have the column header "Group label" while the user guide specifies that it should be "Group Name"
> * RepoConfigParserTest: test config files have the column header "Ignore Commits List" while the user guide specifies that it should be "Ignore Commit List"
> 
> These are relatively small discrepancies. Right now I am using case-insensitive string matching to check the headers, but we can account for these differences quite easily using regex instead. Otherwise, we can just change the headers to match the user guide.
> 
> However, my worry is that users may also be using config files that do not match the user guide specification, which would mean that we would still end up breaking backward compatibility.

Let's make this a separate PR as it is a non-trivial change. Yes, this change itself can be a breaking change, but at least it will make it easier for future changes to be backward compatible.

vig42 replied:
> 
> Let's make this a separate PR as it is a non-trivial change. Yes, this change itself can be a breaking change, but at least it will make it easier for future changes to be backward compatible.

I have made the PR: #1459.

vig42 commented:
> 
> 
> Thanks for the summary @vig42
> I wonder if the flag name is too general e.g., what if we introduce other ways to optimize cloning in future? Perhaps make it more specific? e.g., `--partial-cloning`?

I was worried about that too. I ended up making it general (`--optimize-cloning`) to make it simpler for users to understand, but I think your point is more relevant. I will change it to `--shallow-cloning`, since the main reason for the reduced size of the cloned repo is that it is a shallow clone.

fzdy1914 reviewed (changes requested):
Logic wise, it seems to be OK for me. However, can you add some tests to cover the newly added code?

vig42 commented:
> 
> 
> Logic wise, it seems to be OK for me. However, can you add some tests to cover the newly added code?

I have added JUnit tests (ArgsParserTest, RepoConfigParserTest, RepoConfigurationTest) as well as system tests.

> 
> 
> Thanks for the summary @vig42
> I wonder if the flag name is too general e.g., what if we introduce other ways to optimize cloning in future? Perhaps make it more specific? e.g., `--partial-cloning`?

I have adjusted the flag to `--shallow-cloning`.

> 
> I see. You may want to mention this in the documentation as well.

I have updated the documentation accordingly.

Tejas2805 commented on docs/ug/cli.md:
```diff
@@ -107,6 +107,15 @@ The last modified dates will be in the same timezone specified with the `--timez
 
 <!-- ------------------------------------------------------------------------------------------------------ -->
 
+### `--shallow-cloning`, `-S`
```
This page is arranged alphabetically as of now. So better to push it down accordingly.

Tejas2805 commented on docs/ug/cli.md:
```diff
@@ -107,6 +107,15 @@ The last modified dates will be in the same timezone specified with the `--timez
 
+### `--shallow-cloning`, `-S`
+
+**`--shallow-cloning`**: Clones repositories using Git's shallow cloning functionality, which can significantly reduce the time taken to clone large repositories. However, the flag should not be used for smaller repositories where the .git file is smaller than 500 MB, as it would create overhead. Note: this flag is incompatible with the `--last-modified-date` flag.
```
Instead of `Note:` better to use the `MarkBind info box` we use throughout the page.

Tejas2805 commented on docs/ug/configFiles.md:
```diff
@@ -38,6 +38,7 @@ Ignore Glob List<sup>*+</sup> | The list of file path globs to ignore during ana
 Ignore standalone config | To ignore the standalone config file (if any) in the target repository, enter **`yes`**. If the cell is empty, the standalone config file in the repo (if any) will take precedence over configurations provided in the csv files.
 Ignore Commit List<sup>*+</sup> | The list of commits to ignore during analysis. For accurate results, the commits should be provided with their full hash. Additionally, a range of commits can be specified using the `..` notation, e.g., `abc123..def456` (both inclusive).
 Ignore Authors List<sup>*+</sup> | The list of authors to ignore during analysis. Authors should be specified by their [Git Author Name](#a-note-about-git-author-name).
+Shallow Cloning | Enter **`yes`** to clone the repository using Git's shallow cloning functionality. This option can significantly reduce the time taken to clone large repositories. However, the option should ideally be disabled for smaller repositories, where the .git file is smaller than 500 MB. Note: this is incompatible with the "--last-modified-date" CLI flag.
```
Same suggestion for `Note` here

Tejas2805 commented on src/main/java/reposense/git/GitClone.java:
```diff
@@ -119,4 +169,32 @@ public static void cloneFromBareAndUpdateBranch(Path rootPath, RepoConfiguration
+    }
+
+    /**
+     * Constructs the command to clone a bare repo specified in the {@code config}
```
You have two functions with a duplicate comment header. Better to modify to show what each function exactly does. Will be easier for other devs to understand.

Tejas2805 commented on src/main/java/reposense/git/GitClone.java:
```diff
@@ -119,4 +169,32 @@ public static void cloneFromBareAndUpdateBranch(Path rootPath, RepoConfiguration
+    }
+
+    /**
+     * Constructs the command to clone a bare repo specified in the {@code config}
```
Same suggestion for this.

Tejas2805 reviewed (changes requested):
Left comments regarding some small changes. Overall, great work done!

Just a doubt - What happens if both the CLI arguments are passed - S and last-modified-date?

vig42 commented:
> 
> 
> Left comments regarding some small changes. Overall, great work done!

Have made all the suggested changes.

> 
> Just a doubt - What happens if both the CLI arguments are passed - S and last-modified-date?

Right now it just puts the incorrect "last modified" dates in the JSON. My assumption is that users would read the docs and avoid doing that.

We could make the CLI arguments mutually exclusive so we would not face that conflict. However, we would still face an issue if the "Shallow Clone" option is enabled in the repo config file. One way to handle that could be to modify the FileInfoAnalyzer part here:

https://github.com/reposense/RepoSense/blob/fc288b461f87a14fc7004c0503fe4adbe12360b4/src/main/java/reposense/authorship/FileInfoAnalyzer.java#L106-L112

We could change line 106 to:

```java
if (config.isLastModifiedDateIncluded() && !config.isShallowCloningPerformed()) 
```

So for any repo where the Shallow Cloning config option is enabled, RepoSense would not put the last modified date in the JSON (we could also give the users a warning log if this happens).

Tejas2805 commented:
> Right now it just puts the incorrect "last modified" dates in the JSON. My assumption is that users would read the docs and avoid doing that.

It is possible that the user wants to shallow clone one repo out of say 5. And wants the last-modified-date for all (though we won't get for the one with shallow clone obviously. Therefore, better to generate a warning. 

> So for any repo where the Shallow Cloning config option is enabled, RepoSense would not put the last modified date in the JSON. We could also give the users a warning log if this happens.

I think it is better to give the warning log from our side in case both shallow-cloning and last-modified-date are given? Saying that the last modified dates for this repo will be incorrect or invalid since shallow cloning has been enabled. @damithc your thoughts?

damithc commented:
> I think it is better to give the warning log from our side in case both shallow-cloning and last-modified-date are given? Saying that the last modified dates for this repo will be incorrect or invalid since shallow cloning has been enabled. @damithc your thoughts?

Sure, we can give a warning for that.

vig42 commented:
> 
> 
> > Right now it just puts the incorrect "last modified" dates in the JSON. My assumption is that users would read the docs and avoid doing that.
> 
> It is possible that the user wants to shallow clone one repo out of say 5. And wants the last-modified-date for all (though we won't get for the one with shallow clone obviously. Therefore, better to generate a warning.
> 
> > So for any repo where the Shallow Cloning config option is enabled, RepoSense would not put the last modified date in the JSON. We could also give the users a warning log if this happens.
> 
> I think it is better to give the warning log from our side in case both shallow-cloning and last-modified-date are given? Saying that the last modified dates for this repo will be incorrect or invalid since shallow cloning has been enabled. @damithc your thoughts?

I have added this warning. I have also made the CLI parameters for `--shallow-cloning` and `--last-modified-date` mutually exclusive.

Tejas2805 reviewed (commented):
Looks okay to me. I will need some time to go through the whole code in detail.

fzdy1914 commented on docs/ug/configFiles.md:
```diff
@@ -38,6 +38,11 @@ Ignore Glob List<sup>*+</sup> | The list of file path globs to ignore during ana
 Ignore standalone config | To ignore the standalone config file (if any) in the target repository, enter **`yes`**. If the cell is empty, the standalone config file in the repo (if any) will take precedence over configurations provided in the csv files.
 Ignore Commit List<sup>*+</sup> | The list of commits to ignore during analysis. For accurate results, the commits should be provided with their full hash. Additionally, a range of commits can be specified using the `..` notation, e.g., `abc123..def456` (both inclusive).
 Ignore Authors List<sup>*+</sup> | The list of authors to ignore during analysis. Authors should be specified by their [Git Author Name](#a-note-about-git-author-name).
+Shallow Cloning | Enter **`yes`** to clone the repository using Git's shallow cloning functionality. This option can significantly reduce the time taken to clone large repositories. However, the option should ideally be disabled for smaller repositories where the .git file is smaller than 500 MB, as it would create overhead.
```
Quote .git as `.git` would be better.

vig42 replied:
I have made this change.

fzdy1914 commented on src/main/java/reposense/authorship/FileInfoAnalyzer.java:
```diff
@@ -104,6 +107,10 @@ private static void aggregateBlameAuthorModifiedAndDateInfo(RepoConfiguration co
             }
 
             if (config.isLastModifiedDateIncluded()) {
+                if (config.isShallowCloningPerformed()) {
```
If the two options already mutually exclusive by the argparser, will it still be possible to go to the branch here?

vig42 replied:
It is possible, since the shallow cloning can also be enabled as an option in the repo-config.csv. So as @Tejas2805 had mentioned above, it is possible for a user to specify 1 repo out of 5 to be shallow cloned while also using the `--last-modified-date` flag.

fzdy1914 commented on src/main/java/reposense/parser/RepoConfigCsvParser.java:
```diff
@@ -78,9 +80,18 @@ protected void processLine(List<RepoConfiguration> results, CSVRecord record) th
                     "Ignoring unknown value " + ignoreStandaloneConfig + " in ignore standalone config column.");
         }
 
+        String shallowCloningConfig = get(record, SHALLOW_CLONING_CONFIG_POSITION);
```
If I understand correctly, in your current design, users cannot choose the depth of shallow cloning?

fzdy1914 replied:
We may extend to this shallow cloning argument in a nicer latest-n-commit command if we allow customizing the depth right? Correct me if I am wrong.

vig42 replied:
> 
> 
> If I understand correctly, in your current design, users cannot choose the depth of shallow cloning?

Yes, that is correct. It is designed to replicate the report normally generated by RepoSense, so it just pulls the minimum amount of history possible while still capturing all commits between the "since" and "until" date.

> 
> 
> We may extend to this shallow cloning argument in a nicer latest-n-commit command if we allow customizing the depth right? Correct me if I am wrong.

I feel that this would be a separate feature altogether. This PR is aimed at speeding up the cloning process for large repos by reducing the size of the cloned repo. From the user's POV, the generated report would be identical (well, besides the incompatibility with `--last-modified-date`).

If I understand you correctly, the last-n-commit command would generate a report containing info about the last N commits only, i.e. it would be an alternative to the "since" and "until" dates. We can implement this for the commits report by modifying `GitLog.getWithFiles()` to look at the last N commits rather than the date range. As for the authorship report, we can implement this by just using the N-th latest commit for the `lastCommitHash`.

This means we wouldn't need to use shallow cloning, meaning we can implement the last-n-commits feature while maintaining compatibility with the `--last-modified-date` flag.

fzdy1914 replied:
Sorry, my bad, I previously thought you will only clone depth 1 only for analyze. It is fine in such a case.

fzdy1914 reviewed (commented):
Is this PR conflicting with the multithreaded one? If so, which PR would you like for me to merge first?

Are we lacking system tests for shallow cloning so there are a lot of warning for uncovered line?

vig42 commented:
> 
> 
> Is this PR conflicting with the multithreaded one? If so, which PR would you like for me to merge first?

There are some minor conflicts but they are quite trivial to resolve. They are mainly in the Location/ConfigCliArguments and the ArgsParser, since both PRs make changes to those files. Basically just need to make sure the parameters added by each of the 2 PRs are both included in the merged commit. It shouldn't make a difference which PR is merged first.

> 
> Are we lacking system tests for shallow cloning so there are a lot of warning for uncovered line?

I had added 2 system tests that replicate the existing system tests, except with shallow cloning enabled: `testSinceBeginningDateRangeWithShallowCloning` and `test30DaysFromUntilDateWithShallowCloning`. I feel that this should be sufficient to cover the shallow cloning code in `RepoCloner.java`, so I am not sure why the lines are listed as uncovered.

fzdy1914 commented:
Can you resolve the conflicts and then request a review from the stage-1 active reviewer? Thanks.

Tejas2805 commented on src/main/java/reposense/git/GitCatFile.java:
```diff
@@ -0,0 +1,57 @@
+        String catFileCommand = "git cat-file -p " + commitHash;
+        try {
+            String output = runCommand(rootPath, catFileCommand);
+            List<String> parents = new ArrayList();
```
Can name `parentCommits` instead of `parents`?

vig42 replied:
I have changed this.

Tejas2805 commented on src/main/java/reposense/git/GitCatFile.java:
```diff
@@ -0,0 +1,57 @@
+        List<String> parents = new ArrayList();
+        for (String commitHash : commitHashes) {
+            try {
+                parents.addAll(getParentCommits(root, commitHash));
```
Same here

vig42 replied:
I have changed this.

Tejas2805 commented on src/main/java/reposense/git/GitRevList.java:
```diff
@@ -98,4 +100,24 @@ private static String getAllCommitHashSince(String root, String branchName, Stri
+    /**
+     * Returns true if the repository is empty.
+     */
+    public static boolean getIsEmptyRepo(String root) {
```
Doesn't sound like a boolean returning function with the `get`.  Maybe `check` or `is` will make it sound better.

vig42 replied:
I have changed it to `checkIsEmptyRepo`

Tejas2805 commented on src/main/java/reposense/parser/ArgsParser.java:
```diff
@@ -159,6 +159,15 @@ private static ArgumentParser getArgumentParser() {
                         + "If not provided, default file formats will be used.\n"
                         + "Please refer to userguide for more information.");
 
+        parser.addArgument(TIMEZONE_FLAGS)
```
Any reason for shifting it here? Or just arrangement?

vig42 replied:
This was originally at the bottom after the mutex flags. When I added in the 2 new mutex flags, I moved this above all the mutex flags so it would be a bit clearer.

fzdy1914 replied:
Yes, it will be clearer now.

Tejas2805 commented on src/test/java/reposense/git/GitCatFileTest.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Test(expected = CommitNotFoundException.class)
+    public void getParentCommits_nonexistentCommit_throwsEmptyCommitException() throws Exception {
```
```suggestion
    public void getParentCommits_nonExistentCommit_throwsEmptyCommitException() throws Exception {
```

vig42 replied:
I have changed this.

Tejas2805 commented on src/test/java/reposense/git/GitShowTest.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @Test(expected = CommitNotFoundException.class)
+    public void getCommitDate_nonexistentCommit_throwsEmptyCommitException() throws Exception {
```
```suggestion
    public void getCommitDate_nonExistentCommit_throwsEmptyCommitException() throws Exception {
```

vig42 replied:
I have changed this.

Tejas2805 reviewed (commented):
Left some comments. The code and flow LGTM!

fzdy1914 commented on src/main/java/reposense/report/RepoCloner.java:
```diff
@@ -47,7 +63,41 @@
+            Date sinceDate;
+            try {
+                List<String> distinctParents = graftedCommitParents.stream().distinct().collect(Collectors.toList());
+                sinceDate = GitShow.getEarliestCommitDate(partialBareRoot, distinctParents);
```
Just a question, will just check the date one by one be very time-consuming? Or the number of parents will be very limited given that the date range given is not so much?

vig42 replied:
The number of parents is quite limited. For the cases I have manually checked, the number of grafted commits has been < 5. So even if all of them are merge commits with 2 parents, the number of parents to check is at most 10.

fzdy1914 commented on src/main/java/reposense/util/FileUtil.java:
```diff
@@ -64,7 +66,7 @@ public static void zipFoldersAndFiles(List<Path> pathsToZip, Path sourceAndOutpu
      */
     public static void zipFoldersAndFiles(List<Path> pathsToZip,
-            Path sourcePath, Path outputPath, String... fileTypes) {
+                                          Path sourcePath, Path outputPath, String... fileTypes) {
```
Just a question, does it violate our code standard? I kind of forgot.

vig42 replied:
Oh, good catch. I have fixed the indentation. Not sure why it got changed in the first place.

fzdy1914 reviewed (approved):
Mostly OK for me by just checking the code logic online.

Tejas2805 commented on docs/ug/cli.md:
```diff
@@ -107,7 +107,11 @@ This flag overrides the `Ignore standalone config` field in the CSV config file.
-The last modified dates will be in the same timezone specified with the `--timezone` flag.
+<box type="info" seamless>
+
+* Cannot be used with ``--shallow-cloning`.
```
There are two `` in this line. Needs to be fixed.

vig42 replied:
Fixed

Tejas2805 commented on docs/ug/cli.md:
```diff
@@ -150,6 +154,20 @@ Cannot be used with `--config`.
+
+<box type="info" seamless>
+
+Cannot be used with ``--last-modified-date`.
```
Same here.

vig42 replied:
Fixed

Tejas2805 reviewed (changes requested):
LGTM. Last 2 errors need to be fixed.

Tejas2805 reviewed (approved):
LGTM!
